geo/src/algorithm/sweep/mod.rs

```diff
@@ -1,29 +1,364 @@
+    where
+        F: FnMut(&SweepLineInterval<T>, &SweepLineInterval<T>),
+    {
+        // Sort both by x-coordinate
+        self.insert_events.sort();
+        self.delete_events.sort();
+
+        // Create a lookup array to map segment indices to their DELETE x-coordinates
+        // This avoids O(n) search for each INSERT event
+        // This will let us look up the insert event for a given interval index in constant time
+        // JTS (and GEOS) don't need this as they use object references and pointers respectively
+        // but these approaches are (obviously) a pain in Rust, so we need an intermediate mapping
+        // https://github.com/simplegeo/jts/blob/3a4c9a9c3a7d3274e16bdd2f341df7d3e113d81b/src/com/vividsolutions/jts/index/sweepline/SweepLineIndex.java#L57-L59
+        // We know the number of segments equals the number of INSERT events
+        let num_segments = self.insert_events.len();
+        let mut delete_x_by_index = vec![None; num_segments];
+
+        self.delete_events
+            .iter()
+            .filter(|event| event.interval.index < num_segments)
+            .for_each(|event| {
+                delete_x_by_index[event.interval.index] = Some(event.x_value);
+            });
+
+        // Process INSERT events in sorted order
+        self.insert_events
+            .iter()
+            .enumerate()
+            .for_each(|(i, current_event)| {
+                let current_interval = &current_event.interval;
+                let current_segment_index = current_interval.index;
+
+                // Get the DELETE x-coordinate from our lookup array
+                // Every segment should have both INSERT and DELETE events
+                let delete_x = delete_x_by_index[current_segment_index]
+                    .expect("Every segment should have a DELETE event");
+
+                // Check all INSERT events that start after this one
+                // This iterator chain implements the core sweep line logic:
+                // 1. Start from the next INSERT event (i+1)
+                // 2. Only examine segments that start BEFORE the current segment ends
+                //    (their INSERT x <= our DELETE x), as these are the only ones that
+                //    could potentially intersect with the current segment
+                // 3. Skip self-intersections by filtering out segments with the same index
+                // 4. Check if the x-intervals actually overlap using intervals_overlap
+                // 5. For each overlapping pair, carry out the actual geometric intersection test
+                self.insert_events[i + 1..]
+                    .iter()
+                    .take_while(|other_event| other_event.interval.min <= delete_x)
+                    .filter(|other_event| other_event.interval.index != current_segment_index)
+                    .filter(|other_event| {
+                        intervals_overlap(current_interval, &other_event.interval)
```

Comment:
Small perf change: 

Given that `other_event.interval.min <= delete_x` isn't this check redundant?